Features/config

MaFeLP opened the pull request:
This Pull Request adds a configuration file to configure the bots at runtime.

An example can be found at `<PROJECT ROOT>/config.toml.example`.

The config file configures the bots in the following ways:

- Set a token for each bot (string) [^1]
- Set the automatic responses to messages (table):
    - Trigger (list of strings): One of them needs to be in the message to trigger it
    - Response_pool (list of strings): Selects one random entry, if length is `1`, it therefore just selects this one entry.

An upgrade from `v0.1.2` is easily possible, as it creates a configuration file, if none is found.

[^1]: Warning: A token in the config file will be overridden by the environment variables `DISCORD_TOKEN_XD` or `DISCORD_TOKEN_KAENGURU`!

MaFeLP committed:
Added toml dependency for the config file

Signed-off-by: MaFeLP <60669873+MaFeLP@users.noreply.github.com>

MaFeLP committed:
Updated all dependencies

Signed-off-by: MaFeLP <60669873+MaFeLP@users.noreply.github.com>

MaFeLP committed:
Added config parser

Signed-off-by: MaFeLP <60669873+MaFeLP@users.noreply.github.com>

MaFeLP committed:
Added an example configuration

Signed-off-by: MaFeLP <60669873+MaFeLP@users.noreply.github.com>

MaFeLP committed:
Merge branch 'main' into features/config

MaFeLP committed:
Fixed compiling errors

Signed-off-by: MaFeLP <60669873+MaFeLP@users.noreply.github.com>

MaFeLP committed:
Rereferenced the docstring to the field and not the macro

Signed-off-by: MaFeLP <60669873+MaFeLP@users.noreply.github.com>

MaFeLP committed:
Restructured the configuration file for better extentability

Signed-off-by: MaFeLP <60669873+MaFeLP@users.noreply.github.com>

MaFeLP committed:
Made all fields public

Signed-off-by: MaFeLP <60669873+MaFeLP@users.noreply.github.com>

MaFeLP committed:
Removed unneeded imports

Signed-off-by: MaFeLP <60669873+MaFeLP@users.noreply.github.com>

MaFeLP committed:
Added token config in the config file

Signed-off-by: MaFeLP <60669873+MaFeLP@users.noreply.github.com>

MaFeLP committed:
Made the response Struct clonable

Signed-off-by: MaFeLP <60669873+MaFeLP@users.noreply.github.com>

MaFeLP committed:
Added rand dependency for random numbers

Signed-off-by: MaFeLP <60669873+MaFeLP@users.noreply.github.com>

MaFeLP committed:
Moved autokommentator's replies to the config file

Signed-off-by: MaFeLP <60669873+MaFeLP@users.noreply.github.com>

MaFeLP committed:
Updated example files to include the config file

Signed-off-by: MaFeLP <60669873+MaFeLP@users.noreply.github.com>

MaFeLP committed:
Outsourced reply logic to crate::replies::reply_to

Signed-off-by: MaFeLP <60669873+MaFeLP@users.noreply.github.com>

MaFeLP committed:
reply_to now returns Result

Signed-off-by: MaFeLP <60669873+MaFeLP@users.noreply.github.com>

MaFeLP committed:
Fixed Borrow-checker compiling error

Signed-off-by: MaFeLP <60669873+MaFeLP@users.noreply.github.com>

MaFeLP committed:
Added comment

Signed-off-by: MaFeLP <60669873+MaFeLP@users.noreply.github.com>

MaFeLP committed:
Added documentation to replies::*

Signed-off-by: MaFeLP <60669873+MaFeLP@users.noreply.github.com>

MaFeLP committed:
Refactored some variables, as they ARE in use

- `_data_about_bot` -> `data_about_bot`
- `_ctx` -> `ctx`
- `_new_message` -> `new_message`

Signed-off-by: MaFeLP <60669873+MaFeLP@users.noreply.github.com>

MaFeLP committed:
Removed mutability from new_message

Signed-off-by: MaFeLP <60669873+MaFeLP@users.noreply.github.com>

MaFeLP committed:
Added "normal" reaction to the Känguru Knecht

Signed-off-by: MaFeLP <60669873+MaFeLP@users.noreply.github.com>

MaFeLP committed:
Now handling errors in autokommentator

Signed-off-by: MaFeLP <60669873+MaFeLP@users.noreply.github.com>

MaFeLP committed:
Merge branch 'main' into features/config

Fixes merge conflicts in src/main.rs

MaFeLP committed:
Added documentation to config.rs

Signed-off-by: MaFeLP <60669873+MaFeLP@users.noreply.github.com>

MaFeLP committed:
Added reqwest and futures dependencies

Signed-off-by: MaFeLP <60669873+MaFeLP@users.noreply.github.com>

MaFeLP committed:
Added function to get the default configuration from this repository on GitHub

Signed-off-by: MaFeLP <60669873+MaFeLP@users.noreply.github.com>

MaFeLP committed:
Added documentation to config::make_default_config

Signed-off-by: MaFeLP <60669873+MaFeLP@users.noreply.github.com>

MaFeLP committed:
Switched static config file generation.

This is because reqwest isn't working for me (?), but it also ensures
that the config file is compatible with the current version.

Removed explicit futures and reqwest dependencies

Signed-off-by: MaFeLP <60669873+MaFeLP@users.noreply.github.com>

MaFeLP committed:
Added a more useful error message when the configuration file is invalid.

Signed-off-by: MaFeLP <60669873+MaFeLP@users.noreply.github.com>

MaFeLP committed:
Added a version checker, to check if config version is compatible.

This makes it easier, to distinguish between a wrong config file
and an updated file.

Signed-off-by: MaFeLP <60669873+MaFeLP@users.noreply.github.com>

MaFeLP committed:
Fixed typo

Signed-off-by: MaFeLP <60669873+MaFeLP@users.noreply.github.com>

G0rger commented on Dockerfile:
```diff
@@ -9,4 +9,6 @@ FROM debian:buster-slim
 WORKDIR /app
 #RUN apt-get update && apt-get install -y extra-runtime-dependencies && rm -rf /var/lib/apt/lists/*
 COPY --from=builder /xd_bot /app/xd_bot
+VOLUME ./config.toml /app/config.toml
```
This should be (per the spec at https://docs.docker.com/engine/reference/builder/#volume)
```suggestion
VOLUME /app/config.toml
```

G0rger commented on src/config.rs:
```diff
@@ -0,0 +1,233 @@
+    /// The token that is used to log into discord
+    pub token: Option<String>,
+    /// The replies and messages that this bot should react to.
+    pub replies: Vec<Response>
```
I think it would be more consistent if this field was named `responses`.
```suggestion
    pub responses: Vec<Response>
```

lxgr-linux replied:
+1

MaFeLP replied:
9d83b09daca5fd6b7e3892f83d855f06bb94df03: Hope you are happy

G0rger commented on src/config.rs:
```diff
@@ -0,0 +1,233 @@
+    /// The token that is used to log into discord
+    pub token: Option<String>,
+    /// The replies and messages that this bot should react to.
+    pub replies: Vec<Response>
```
Same here
```suggestion
    pub responses: Vec<Response>
```

G0rger commented on src/config.rs:
```diff
@@ -0,0 +1,233 @@
+    pub trigger: value::Array,
+    /// A list of strings that are replied to the message:
+    /// If there are multiple elements in this list, one is selected randomly.
+    pub response: value::Array,
```
Shouldn't this be plural?
```suggestion
    pub responses: value::Array,
```

G0rger commented on src/config.rs:
```diff
@@ -0,0 +1,233 @@
+    fn clone(&self) -> Self {
+        Response {
+            trigger: self.trigger.to_vec(),
+            response: self.response.to_vec()
```
```suggestion
            responses: self.responses.to_vec()
```

G0rger commented on src/config.rs:
```diff
@@ -0,0 +1,233 @@
+        let mut out = String::from("0.0");
+        let version_line = Regex::new("^version *= *\"\\d\\.\\d\" *(|#.*)$").unwrap();
+        let version_matcher = Regex::new("(\\d*\\.\\d*)").unwrap();
+        for line in config_content.lines() {
```
This is not how you do regex!
I will do another PR later for that though.

MaFeLP replied:
Does 1d73d5bf9172d34afd6a8f082411544279039e04 fix this issue for you?

G0rger replied:
Not fully

G0rger replied:
I will give you a patch when I'm finished

G0rger replied:
Marking this as resolved for the other two comments

G0rger commented on src/main.rs:
```diff
@@ -3,26 +3,58 @@ extern crate lazy_static;
+        Err(_) => {
+            let config_arc = Arc::clone(&CONFIG);
+            let mut config_lock = config_arc.lock();
+            while config_lock.is_err() {
```
A error in locking means the lock has been poisoned and can't be unlocked again.
This should panic instead of retrying!

G0rger commented on src/config.rs:
```diff
@@ -0,0 +1,233 @@
+    ///
+    /// let config_arc = Arc::clone(&CONFIG);
+    /// let mut config_lock = config_arc.lock();
+    /// while config_lock.is_err() {
```
A error in locking means the lock has been poisoned and can't be unlocked again.
This should panic instead of retrying!

G0rger commented on src/main.rs:
```diff
@@ -38,12 +70,37 @@ async fn start_xd() {
+        Err(_) => {
+            let config_arc = Arc::clone(&CONFIG);
+            let mut config_lock = config_arc.lock();
+            while config_lock.is_err() {
```
A error in locking means the lock has been poisoned and can't be unlocked again.
This should panic instead of retrying!

lxgr-linux commented on src/config.rs:
```diff
@@ -0,0 +1,233 @@
+use std::{
+    fs,
+    env,
+    path::Path,
+    sync::{Arc, Mutex},
+};
+use std::fs::File;
+use std::io::Write;
+use std::process::exit;
```
```suggestion
use std::{
    fs::{self, File},
    env,
    path::Path,
    io::Write,
    process::exit,
    sync::{Arc, Mutex},
};
```
Why not like this?

lxgr-linux commented on src/kaenguru.rs:
```diff
@@ -6,8 +6,9 @@ use serenity::{
 };
 use serenity::utils::Color;
-use crate::kaenguru;
+use crate::config::Bots;
 use crate::kaenguru::euro_to_mark::get_euro;
+use crate::replies::{reply_to, ReplyError};
```
Why the full path?

MaFeLP replied:
Because the project does not compile with `crate::` or `crate::replies::` removed.

lxgr-linux commented on src/replies.rs:
```diff
@@ -0,0 +1,153 @@
+use std::sync::Arc;
+use std::thread::sleep;
+use std::time::Duration;
```
```suggestion
use std::{
    sync::Arc,
    thread::sleep,
    time::Duration,
};
```

MaFeLP committed:
Update Dockerfile

Co-authored-by: Jannis Fricke <jannis@gorger.dev>

MaFeLP committed:
Reformatted uses

Co-authored-by: lxgr-linux <64218491+lxgr-linux@users.noreply.github.com>

MaFeLP committed:
Removed waiting on Mutex Lock Error to make @G0rger happy

Signed-off-by: MaFeLP <60669873+MaFeLP@users.noreply.github.com>

Co-authored-by: Jannis Fricke <jannis@gorger.dev>

MaFeLP committed:
Fixed compiler warnings

Signed-off-by: MaFeLP <60669873+MaFeLP@users.noreply.github.com>

MaFeLP committed:
Did some refactoring

- Kaenguru::replies -> Kaenguru::responses
- Autokommentator::replies -> Autokommentator::responses
- Response::response -> Response::response_pool

Signed-off-by: MaFeLP <60669873+MaFeLP@users.noreply.github.com>

Co-authored-by: Jannis Fricke <jannis@gorger.dev>

G0rger commented on src/config.rs:
```diff
@@ -25,22 +24,16 @@ lazy_static! {
-    ///         println!("Something went wrong internally...");
-    ///         return
+    ///     Err(why) => {
+    ///         eprintln!("Something went wrong internally: {:?}\nMutex is poisoned: {}", why, why);
+    ///         return;
```
This should panic instead of just returning.
A Mutex is poisened when the thread that has it locked panics. It becomes unusable afterwards.
```suggestion
    ///         panic!("Something went wrong internally: {:?}\nMutex is poisoned: {}", why, why);
```

MaFeLP replied:
0911b0d405d1fedbb5d8e5221c53f3fa56a5ebc7

G0rger commented on src/main.rs:
```diff
@@ -45,8 +40,8 @@ async fn start_xd() {
-                    println!("Something went wrong internally...");
+                Err(why) => {
+                    eprintln!("Something went wrong internally: {:?}\nMutex is poisoned: {}", why, why);
                     return
```
Panic!
```suggestion
                    panic!("Something went wrong internally: {:?}\nMutex is poisoned: {}", why, why);
```

MaFeLP replied:
0911b0d405d1fedbb5d8e5221c53f3fa56a5ebc7

G0rger commented on src/replies.rs:
```diff
@@ -54,22 +54,17 @@ pub enum ReplyError {
-                println!("Something went wrong internally...");
+            Err(why) => {
+                eprintln!("Something went wrong internally: {:?}\nMutex is poisoned: {}", why, why);
                 return Err(ReplyError::MutexAcquisition);
```
Panic!
```suggestion
                panic!("Something went wrong internally: {:?}\nMutex is poisoned: {}", why, why);
```

MaFeLP replied:
0911b0d405d1fedbb5d8e5221c53f3fa56a5ebc7

MaFeLP committed:
Simplified regex to make @G0rger happy

Reference: https://github.com/MaFeLP/discord_bots/pull/8#discussion_r775621025

Signed-off-by: MaFeLP <60669873+MaFeLP@users.noreply.github.com>

Co-authored-by: Jannis Fricke <jannis@gorger.dev>

MaFeLP committed:
Made the thread panic after mutex poisoning instead of returning an error.

Signed-off-by: MaFeLP <60669873+MaFeLP@users.noreply.github.com>

Co-authored-by: Jannis Fricke <jannis@gorger.dev>

MaFeLP committed:
Bumped package version to 0.2.0

Signed-off-by: MaFeLP <60669873+MaFeLP@users.noreply.github.com>

MaFeLP committed:
Fixed compiler warnings

Signed-off-by: MaFeLP <60669873+MaFeLP@users.noreply.github.com>

G0rger commented on src/config.rs:
````diff
@@ -0,0 +1,219 @@
+    /// };
+    /// ```
+    ///
+    pub static ref CONFIG: Arc<Mutex<Config>> = Arc::new(Mutex::new(Config::new()));
````
Put the regex always in a lazy_static
```suggestion
    pub static ref CONFIG: Arc<Mutex<Config>> = Arc::new(Mutex::new(Config::new()));

    static ref VERSION_REGEX: Regex = Regex::new("(?m)^version *= *\"(?P<version>\\d*\\.\\d*)\" *(|#.*)$").unwrap();
```

G0rger commented on src/config.rs:
````diff
@@ -0,0 +1,219 @@
+///     panic!("Config version incompatible!");
+/// }
+/// ```
+fn check_version(config_content: &String) -> (bool, String) {
+    // Get the version, by looping over the config contents, searching for the config line and
+    // getting the version part of it.
+    let version = {
+        let version_line = Regex::new("(?m)^version *= *\"(?P<major>\\d*)\\.(?P<minor>\\d*)\" *(|#.*)$").unwrap();
+        let captures = version_line.captures(config_content).unwrap();
+        let out = format!("{}.{}", captures.name("major").unwrap().as_str(), captures.name("minor").unwrap().as_str());
+        dbg!("{}", &out);
+        out
+    };
+    // Config file is always compatible with its associated program version
+    if format!("{}.{}", env!("CARGO_PKG_VERSION_MAJOR"), env!("CARGO_PKG_VERSION_MINOR")).eq_ignore_ascii_case(&version) {
+        return (true, version);
+    }
+    // If config file version is not the program version, check if it is still compatible:
+    for (major, minor) in COMPATIBLE_VERSIONS {
+        if format!("{}.{}", major, minor).eq_ignore_ascii_case(&version) {
+            return (true, version);
+        }
+    }
+    return (false, version);
+}
````
Use single matching group, accept any string slice, return string slice
```suggestion
fn check_version(config_content: &str) -> (bool, &str) {
    // Get the version, by searching for the config line and getting the version part of it.
    let version = {
        let captures = VERSION_REGEX.captures(config_content).unwrap();
        let out = captures.name("version").unwrap().as_str();
        dbg!("{}", out);
        out
    };
    // Config file is always compatible with its associated program version
    if format!("{}.{}", env!("CARGO_PKG_VERSION_MAJOR"), env!("CARGO_PKG_VERSION_MINOR")).eq_ignore_ascii_case(version) {
        return (true, version);
    }
    // If config file version is not the program version, check if it is still compatible:
    for (major, minor) in COMPATIBLE_VERSIONS {
        if format!("{}.{}", major, minor).eq_ignore_ascii_case(version) {
            return (true, version);
        }
    }
    return (false, version);
}
```

G0rger reviewed (changes requested):
Otherwise good!

MaFeLP committed:
Made version checking faster

- Now static to only compile once
- Checks for the version as a whole and not individual MAJOR & MINOR versions

Co-authored-by: Jannis Fricke <jannis@gorger.dev>

lxgr-linux reviewed (approved):
LGTM

G0rger reviewed (approved):
LGTM

MaFeLP committed:
Made panic message clearer, when no version was found in the config file

Signed-off-by: MaFeLP <60669873+MaFeLP@users.noreply.github.com>
